app/models/fs/message.rb

```diff
@@ -0,0 +1,133 @@
+class Fs::Message
+  FS_SUBJECT_NAME = 'Finančná správa'
+
+  def self.create_inbox_message_with_thread!(raw_message, box:)
+    message = nil
+    associated_outbox_message = box.messages.where("metadata ->> 'fs_message_id' = ?", raw_message['sent_message_id']).take
+
+    MessageThread.with_advisory_lock!(associated_outbox_message.metadata['correlation_id'], transaction: true, timeout_seconds: 10) do
+      message = create_inbox_message(raw_message)
+
+      message.thread = associated_outbox_message.thread
+
+      message.save!
+    end
+
+    create_message_objects(message, raw_message)
+    update_html_visualization(message)
+
+    EventBus.publish(:message_created, message)
+  end
+
+  def self.create_outbox_message_with_thread!(raw_message, box:)
+    message = nil
+    associated_message_draft = box.messages.where(type: 'Fs::MessageDraft').where("metadata ->> 'fs_message_id' = ?", raw_message['message_id']).take
+
+    merge_identifier = (associated_message_draft.metadata['correlation_id'] if associated_message_draft) || SecureRandom.uuid
+
+    MessageThread.with_advisory_lock!(merge_identifier, transaction: true, timeout_seconds: 10) do
+      message = create_outbox_message(raw_message)
+
+      message.thread = box.message_threads.find_or_create_by_merge_uuid!(
+        box: box,
+        merge_uuid: merge_identifier,
+        title: message.title,
+        delivered_at: message.delivered_at
+      )
+
+      associated_message_draft.destroy if associated_message_draft
```

Comment:
Nebude tu rovnaky problem so stracajucimi sa tagmi?

Reply:
Bude, myslim na to, ale logiku toho by som vyladila v tom otvorenom PR najprv a potom to lahko preklopim aj sem. Nech to neriesime teraz na dvoch miestach.